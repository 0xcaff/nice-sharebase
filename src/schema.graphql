type Library {
	id: ID!
	name: String!
	isPrivate: Boolean!

	folders: [Folder]!
}

type Folder {
	id: ID!
	name: String!

	folders: [Folder]!
	documents: [Document]!
}

# If you want to access the contents of a document or create a new document,
# you will need to talk to the orignial ShareBase server. Checkout description
# of the Query type fore more.

type Document {
	id: ID!
	name: String!

	# The epoch time at which this document was last modified.
	modified: String!
}

# There are two parts to the ShareBase public API: the data storage and retrival
# part and the metadata management part. This API exposes the metadata part
# through a GraphQL interface. If you want to access the data storage part, talk
# directly to the official ShareBase API. The documentation for the official
# ShareBase API can be found at https://developers.sharebase.com/
#
# Authorization is handled by passing a token to the HTTP Authorization header
# of a request. You can get a token by calling the authenticate mutation. If you
# want to use the original ShareBase API for authentication, you can pass tokens
# using the PHOENIX-TOKEN scheme to this header also. All endpoints except the
# authenticate mutation require authentication.
type Query {
	me: Authed!
	libraries: [Library]!
	library(id: ID!): Library!
	folder(id: ID!): Folder!
	document(id: ID!): Document!
}

type Authed {
	# The access token. Specify this string in the token property to
	# createContext prefixed with BOX-TOKEN like so:
	#
	#   BOX-TOKEN THIS-IS-MY-TOKEN
	#
	# and you will be able to access protected resources on the server.
	token: ID!

	# The name of the currently signed in user.
	name: String!

	# The id of the currently signed in user.
	id: ID!
}

type PNXToken {
	# The phoenix token used to authenticate directly to the offical ShareBase
	# servers.
	token: ID!

	# The epoch time at which this token will no longer be valid.
	expires: Int!
}

type Share {
	id: ID!
	reference: String!

	# A url which can be used to access the shared resource.
	url: String!
}

# Setting passed when sharing folders and documents. If expiresOn and singleUse
# are unset, the share will never expire. If more than one is set, an exception
# is thrown. If none are set, the link share will never expire.
input ShareInput {
	# The epoch time at which this share will expire.
	expiresOn: Int
	singleUse: Boolean

	# A password which will be required when acessing the share. This password
	# MUST adhere to the sharebase instance's password policy. If this isn't
	# provided, a password will not be set.
	password: String

	allowView: Boolean!
	allowDownload: Boolean!
	allowUpload: Boolean!
}

type Mutation {
	# Registers your credentials with the server and creates a new session if
	# the provided credentials are valid. Credentials will be held on the server
	# until all the user's sessions are revoked. The token in the response
	# allows you to make authenticated requests with this API until the session
	# is revoked.
	authenticate(email: String!, password: String!): Authed!

	# Revokes the access token being used to make this request. If the session
	# doesn't exist, an error is thrown.
	revokeSession: Boolean!

	# Creates a new phoenix token for direct communication with the orignal
	# ShareBase API. Be careful with this token because it cannot be revoked. It
	# expires in 10 hours.
	phoenixToken: PNXToken!

	# Creates a new folder in the library at the specified path. Required parent
	# folders are created if they don't exist. The path delimiters aren't
	# specified. Try whatever you want.
	newFolder(libraryId: ID!, path: String!): Folder!
	deleteFolder(id: ID!): Boolean!
	shareFolder(id: ID!, input: ShareInput!): Share!
	revokeFolderShare(folderId: ID!, shareId: ID!): Boolean!

	shareDocument(id: ID!, input: ShareInput!): Share!
	revokeDocumentShare(documentId: ID!, shareId: ID!): Boolean!
	deleteDocument(id: ID!): Boolean!
}
