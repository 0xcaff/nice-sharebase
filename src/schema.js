import { makeExecutableSchema } from 'graphql-tools';

import { resolvers } from './resolvers';
import { create } from './loaders';
import { nop, required } from './utils';
import { initNetwork } from './network';
import rawSchema from './schema.graphql';

export const schema = makeExecutableSchema({typeDefs: [rawSchema], resolvers});

// A to create the context for the nice-sharebase schema. The context holds the
// sessions and loaders responsible for making request sto the ShareBase API.
export const createContext = ({
  // The base url of the official ShareBase API we will be resolving request
  // against.
  base = required('base', 'createContext()'),

  // A function which takes in a Request, modifies it and returns the request to
  // send to the official ShareBase API. transform will always only be called
  // before a request is made to the official ShareBase API.
  transform = nop,

  // A place to keep newly created sessions and retrive informatino about
  // existing sessions. This is used by the client to the original ShareBase API
  // to maintain active sessions. This can be any map like object (set, get,
  // has, clear, delete, values). This should be persisted between requests.
  sessionStore = required('sessionStore', 'createContext()'),

  // A token to use to authenticate against the endpoint. This can be a
  // BOX-TOKEN or PHONIEX-TOKEN. See the Query type for more information about
  // authentication.
  authorization,
}) => {
  const { session: me, token } = tryGetSession(sessionStore, authorization);
  const session = { me, store: sessionStore, token };

  // Information about the requests which were made to the official ShareBase
  // API to fulfill the query / mutation.
  const logs = [];

  const network = initNetwork({ base, transform, logs, session });
  const loaders = create({ network });

  return { loaders, network, session, logs };
};

// Tries to get a session from the given information. If there isn't enough
// information or a session doesn't exist returns a falsy value.
function tryGetSession(store, header) {
  if (!header) {
    return {};
  }

  const [ prefix, creds ] = header.split(' ');
  if (prefix === 'BOX-TOKEN') {
    // try to look up session information
    const session = store.get(creds);

    return { session, token: creds };
  } else if (prefix === 'PHOENIX-TOKEN') {

    return { session: null, token: creds };
  } else {

    return {};
  }
}
