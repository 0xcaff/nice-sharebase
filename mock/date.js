// Sets any calls to date to return a constant time with a lot of zeros.
const constantDate = new Date(1500000000000);

global.Date.now = () => constantDate;