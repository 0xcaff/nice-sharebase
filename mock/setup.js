// Sets up nock to respond to requests to the mock endpoint.
import nock from 'nock';
import path from 'path';
import { readFile, first } from './utils';

export const ENDPOINT = 'http://mock.com/';

const setup = app => app
	.get(_ => true)
	.reply(function(url, body, cb) {
		handle.bind(this)(url, body)
			.then((...args) => {
				// setup for next request
				setup(app);
				cb(null, ...args);
			})
			.catch(err => cb(err));
	});

const app = nock(ENDPOINT);
setup(app);

async function handle(p, reqBody) {
	// get path and trim leading /
	p = p.slice(1);

	const paths = [`${p}.json`, `${p}/index.json`]
 		.map(raw => path.resolve(__dirname, raw));

	const result = await first(paths, path => readFile(path));

	if (result !== undefined) {
			return [200, result, {'Content-Type': 'application/json'}];
	} else {
		const message = `Path not found for ${p}. Tried:`;
		const tried = paths.map(path => '\t' + path).join('\n');
		return [404, message + '\n' + tried];
	}
};

