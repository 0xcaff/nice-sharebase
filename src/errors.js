import { GraphQLError } from 'graphql';

// An error which occurs when making a request fails.
export class FetchError extends GraphQLError {
  constructor(resp, body) {
    super(`Request Failed. Status: ${resp.status} - ${resp.statusText}. Body: ${body}`);

    this.body = body;
    this.resp = resp;
  }
}

export const throwOnFail = async (resp) => {
  if (resp.ok) {
    return resp;
  } else {
    throw new FetchError(resp, await resp.text());
  }
}