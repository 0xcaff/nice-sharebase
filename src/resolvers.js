import { getters } from './utils';
import { auth } from './mutations';

export const resolvers = {
  Query: {
    libraries: (obj, args, { loaders: { library } }) =>
      library.load(undefined, 'all'),

    library: (obj, args, { loaders: { library } }) =>
      library.load(args['id']),

    folder: (obj, args, { loaders: { folder } }) =>
      folder.load(args['id']),

    document: (obj, args, { loaders: { document } }) =>
      document.load(args['id']),

    me: (obj, args, { session: { me: { authed } } }) =>
      authed
  },

  Mutation: {
    authenticate: (obj, args, context) =>
      auth(args['email'], args['password'], context),

    // TODO: implement the rest of the mutations
  },

  Authed: getters({
    token: 'delegate',
    name: 'UserName',
    id: 'UserId',
  }),

  Library: getters({
    id: 'LibraryId',
    name: 'LibraryName',
    isPrivate: 'IsPrivate',

    folders: (obj, args, { loaders: { libraryFolders } }) =>
      libraryFolders.load(obj['LibraryId']),
  }),

  Folder: getters({
    id: 'FolderId',
    name: 'FolderName',

    folders: async (obj, args, { loaders: { folder } }) => {
      const f = await folder.load(obj['FolderId'], { folders: true });
      return f['Embedded']['Folders'];
    },

    documents: async (obj, args, { loaders: { folder } }) => {
      const f = await folder.load(obj['FolderId'], { documents: true });
      return f['Embedded']['Documents'];
    },
  }),

  Document: getters({
    id: 'DocumentId',
    name: 'DocumentName',
    modified: 'DateModified',
  }),
};