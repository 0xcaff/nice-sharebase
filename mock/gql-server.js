import gql from 'express-graphql';
import express from 'express';
import url from 'url';
import { schema, createContext } from '../index.js';

import { ENDPOINT, logs } from './setup';

const app = express();

app.use('/graphql', gql(
  async (req, res, params) => {
    const base = url.resolve(ENDPOINT, 'api/');
    const context = createContext({ base });

    return {
      schema,
      context,
      extensions: (document, variables, operationName, result) => {
        const l = logs.slice();
        logs.length = 0;
        return l;
      },
      graphiql: true,
      pretty: true,
    };
  }
));

app.listen(3010, '0.0.0.0', _ => console.log('listening on 0.0.0.0:3010'));
