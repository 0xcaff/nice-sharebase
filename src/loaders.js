import url from 'url';
import 'isomorphic-fetch';

import { throwOnFail } from './errors';

// Called usually once per request to create the caches responsible for holding
// on to data.
export const create = ({ base, transform = nop}) => ({
  library: LibraryLoaderFor({ base, transform }),
});

const LibraryLoaderFor = ({ base, transform }) => new Loader({
  fetchFn: async (id, extras) => {
    const endpoint = id === 'all' ? 'libraries' : `libraries/${id}`;
    const path = url.resolve(base, endpoint);

    const req = transform(new Request(path));
    const resp = await throwOnFail(await fetch(req));
    const body = await resp.json();

    return body;
  },
  putCacheFn(id, extras, ret) {
    // put all values in cache
    this.cache.set(id, ret);

    if (id === 'all') {
      // add each lib to cache
      ret.forEach(lib => this.cache.set(+lib['LibraryId'], lib));
    };
  },
});

// A minimal data loader which doesn't batch requests since we can gain
// flexibility without it and and we don't need it because the ShareBase API
// doesn't do batching.
class Loader {
  constructor({ fetchFn, tryCacheFn = this.tryCache, putCacheFn = this.putCache, cache }) {
    this.cache = cache || new Map();

    this.tryCache = tryCacheFn.bind(this);
    this.putCache = putCacheFn.bind(this);
    this.fetch = fetchFn.bind(this);
  }

  tryCache(id, extras) {
    return this.cache.get(id);
  }

  putCache(id, extras, thing) {
    this.cache.set(id, thing);
  }

  async load(id, extras) {
    const cached = this.tryCache(id, extras);
    if (cached) {
      return cached;
    }

    const ret = await this.fetch(id, extras);
    this.putCache(id, extras, ret);
    return ret;
  }
}

const nop = (arg) => arg;
