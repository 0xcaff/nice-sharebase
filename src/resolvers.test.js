import url from 'url';
import { graphql } from 'graphql';

import { schema } from './schema';
import { create } from './loaders';
import { logs, ENDPOINT } from '../mock/setup';

import { getters } from './resolvers';
import { FetchError } from './errors';

const getContext = _ => ({
  loaders: create({ base: url.resolve(ENDPOINT, 'api/') }),
});

it('should create getters', () => {
  const data = {
    agent: 820,
    operation: 'Dark Night',
  };

  const acc = getters({
    id: 'agent',
    op: 'operation',
    realFunc: (obj) => 420,
  });

  expect(acc.id(data)).toBe(data['agent']);
  expect(acc.op(data)).toBe(data['operation']);
  expect(acc.realFunc(data)).toBe(420);
});

it('should get all library names', async () => {
  const context = getContext();
  const resp = await graphql(schema, "{ libraries { name } }", undefined, context);

  expect(resp.errors).toBeFalsy();
  expect(resp.data).toMatchSnapshot();
  expect(logs).toMatchSnapshot();
});

it('should fail explicitly when fetching a non-existing resource', async () => {
  const context = getContext();
  const resp = await graphql(schema, "{ library(id: 42069) { name } }", undefined, context);

  expect(resp.errors).toMatchSnapshot();
  expect(logs).toMatchSnapshot();
});

it('should only make the required requests on complex queries', async () => {
  const context = getContext();
  const resp = await graphql(schema, `{ library(id: 406) { folders { name folders { name } documents { name } } } }`, undefined, context);

  expect(resp.errors).toBeFalsy();
  expect(resp.data).toMatchSnapshot();
  expect(logs).toMatchSnapshot();
});
