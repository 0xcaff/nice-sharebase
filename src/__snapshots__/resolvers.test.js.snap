// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fail explicitly when fetching a non-existing resource 1`] = `
Array [
  [GraphQLError: Request Failed. Status: 404 - Not Found. Body: Path not found for api/libraries/42069. Tried:
	/home/drone/projects/nice-sharebase/mock/api/libraries/42069.json
	/home/drone/projects/nice-sharebase/mock/api/libraries/42069/index.json],
]
`;

exports[`should get all library names 1`] = `
Object {
  "libraries": Array [
    Object {
      "name": "My Library",
    },
    Object {
      "name": "Iron Man Suit",
    },
  ],
}
`;
